rxjs-1/src/race-calcs.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+
+const { fromEvent, combineLatest } = require('rxjs');
+const { map, filter, takeUntil, throttleTime, pairwise } = require('rxjs/operators');
+
+/**
+ * Calculate real-time speed of a car
+ * @param {EventEmitter} race 
+ * @param {string} reqCarName 
+ */
+function getCarSpeed(race, reqCarName) {
+  let endInfo$ = fromEvent(race, 'end');
+  return fromEvent(race, 'data').pipe(
+    // Keep Only requested car by name
+    filter(({ carName }) => carName === reqCarName),
+    // Filter bad value when time=0 preventing NaN
+    filter(({ time }) => time != 0),
+    // get the previous value with the current in an array
+    pairwise(),
+    map(function ([prev, current]) { // time in ms, xlocation in m
```

Comment:
From what I understand, you are using 2 successive data (`pairwise`) to calculate the speed. Is that correct?
_Using marble syntax:_
```
-a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emited every 50ms)
---B-C-D-E-F-G-H-I-J-K-L-  <-where B is the speed using `a` and `b`, and C etc..
---B-----E-----H-----K---  <-after thottle of 200ms
```

The guidelines would rather suggest (maybe not that clearly :) ) that you use 2 data points separated from 200ms to calculate the speed.

```
-a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emited every 50ms)

-------D-E-F-G-H-I-J-K-L-  <-where D is the speed calculated using `a` and `d`
                                   E is the speed calculated using `b` and `e`
                                    etc..
```

Do you see a way to do that?

Alternatively, I think that, with your current implementation, if you invert 2 lines you can end up with something similar:
```
-a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emited every 50ms)

-------D-----G-----J-----  <-where D is the speed calculated using `a` and `d`
                                   G is the speed calculated using `d` and `g`
```

Reply:
> From what I understand, you are using 2 successive data (`pairwise`) to calculate the speed. Is that correct?
> _Using marble syntax:_
> 
> ```
> -a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emited every 50ms)
> ---B-C-D-E-F-G-H-I-J-K-L-  <-where B is the speed using `a` and `b`, and C etc..
> ---B-----E-----H-----K---  <-after thottle of 200ms
> ```

Ok I understand that : in fact my throttle is **working on the event listener** time spent to send data, & **not the time of the cars** so throttle is useless in my case & I loose some data !
  
> The guidelines would rather suggest (maybe not that clearly :) ) that you use 2 data points separated from 200ms to calculate the speed.
> 
> ```
> -a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emited every 50ms)
> 
> -------D-E-F-G-H-I-J-K-L-  <-where D is the speed calculated using `a` and `d`
>                                    E is the speed calculated using `b` and `e`
>                                     etc..
> ```
> 
> Do you see a way to do that?

I think that it's possible to do:

- Remove throttle prevent losing some data received
- Add a global variable call it globTime & increment it with the time received
- Calculate globTime+=time 
-  If `globTime < 200`  **just do nothing**
- _If `globTime+=time >= 200` change _ reset `globTime = globTime-200` <== this is the moment I should do something with data

Reply:
[`window`](https://www.learnrxjs.io/operators/transformation/window.html) operator could be useful :)

If you throttle before pairwise, you might be able to get something working similar to what you describe without using global variable (that is not very rxjs ;) )
```
-a-b-c-d-e-f-g-h-i-j-k-l-  <-data (consider emtited every 50ms)
-A-----D-----G-----J-----  <-after throttle of 200ms
-------o-----o-----o-----  <-after with pairwise
    [A,D]  [D,G]  [G,J]
```